### DISCUSSION TITLE
Why almost every object is an interface (e.g. `IXLCell`)?

### DISCUSSION THREAD

**Discussion by VladiStep** [ClosedXML/ClosedXML#2392] [General]
The question is in the title.

**Comment by igitur** [comment]
From my side the answer is... we inherited the project like this from the original author and continued the pattern.

**Reply by VladiStep** [reply]
Thanks for the answer.
But could you also answer why I can't find the implementation for most of the methods?
For example `IXLRow.RowNumber()`.

**Reply by igitur** [reply]
The implementation is at https://github.com/ClosedXML/ClosedXML/blob/ef915a5db05a0a57499e34e418096228f097ffee/ClosedXML/Excel/Ranges/XLRangeBase.cs#L779

**Reply by VladiStep** [reply]
Oh, okay.
So, if one wants to "Go to Definition" of `IXLRow.RowNumber()`, then one should temporary erase the first `I` so it turns into `XLRow.RowNumber()`, and now one can use "Go to Definition" there.

**Reply by vbjay** [reply]
Go to implementation is an option.

**Comment by StevenHughesGWT** [comment]
I cannot speak for the authors of those interfaces/classes, but in general this is done for reasons of Object-Oriented Design.  Dependency Inversion and the Open/Closed Principle.  High-level code should depend upon interfaces, not on low level code/classes (Dependency Inversion). That way, it is easier to extend functionality of a lower-level class (Open/Closed Principle).
